### PR TITLE
Interface between Board and Algorithms

### PR DESCRIPTION
**Pull request by jzych**
Implementation proposition to create interface layer between Board to solve and Algorithms to do so.
Note: Not all code could be tested when no full algorithms are implemented,

### CODE REVIEW

File: test/LogicalImagesTests.cpp [hunk 1]
```diff
@@ -1,6 +1,18 @@
 #include <gtest/gtest.h>
 #include "json.hpp"
 #include "LogicalImages.hpp"
+#include "JSONFileReader.hpp"
+
+class AlgorithmA : public Algorithm
```
Comment (LordLukin) [inline test/LogicalImagesTests.cpp]: I propose an AlgorithmStub name. It should also be in a separate hpp file (no need to have cpp)

File: test/LogicalImagesTests.cpp [hunk 2]
```diff
@@ -15,3 +27,55 @@ TEST_F(LogicalImagesTests, assertThatJSONParsingLibraryWorks)
...
+
+TEST_F(LogicalImagesTests, checkSolveAlgorithmA)
+{
+    std::string path = "../test/ReadMeBoardTestFile.json";
+    std::string solution = "Puzzle solved!";
+    auto a = std::make_shared<AlgorithmA>();
+    testing::internal::CaptureStdout();
+
+    LogicalImages li(path, a);
+    li.solve();
```
Comment (LordLukin) [inline test/LogicalImagesTests.cpp]: Using cout in tests is not advised. You can return some value from solve method and check it.

Comment (jzych) [reply test/LogicalImagesTests.cpp]: How then solve problem of code coverage with tests? Or can't it be used to suppress error msgs from exceptions, so they won't appear on screen when testing?
This functions is purly to run solving algorithm and print solution if found.

File: test/LogicalImagesTests.cpp (hunk 1 above)
Comment (jzych) [reply test/LogicalImagesTests.cpp]: Will move to AlgorthmStub.hpp and also create there AlgorithmStubB, which returns false from solve function.

File: test/LogicalImagesTests.cpp (hunk 2 above)
Comment (LordLukin) [reply test/LogicalImagesTests.cpp]: You can ignore coverage for now if you know that there will be some tests implemented later, which will cover this implementation. 
Coverage is useful, but do not blindly try to achieve 100%. Usually, it means that everything is done for statistics, not real work ;)

File: test/AlgorithmStub.hpp [hunk 3]
```diff
@@ -0,0 +1,23 @@
+#include "Algorithm.hpp"
+
+class AlgorithmStubA : public Algorithm
+{
+public:
+    AlgorithmStubA() {}
```
Comment (LordLukin) [inline test/AlgorithmStub.hpp]: no need to explicitly write default constructor here :) 
Less code = less places for mistakes = less maintenance = better code

Comment (jzych) [reply test/AlgorithmStub.hpp]: done

File: src/Algorithm.hpp [hunk 4]
```diff
@@ -0,0 +1,11 @@
+#pragma once
+#include "Board.hpp"
+
+class Algorithm
+{
+public:
+    Algorithm();
```
Comment (LordLukin) [inline src/Algorithm.hpp]: Not needed

**Pull request merged**